forem/forem
# Navigation Items and Role access for Data Update Scripts

## What type of PR is this? (check all applicable)


## Description
In this PR, the following changes are made to the Data Update Scripts: 
- We keep the route `/admin/data_update_scripts` namespaced under admin and allow access to a single_resource_admin role with resource type DataUpdateScript. 
- Add a nav item under a section called Tech Resources on the /admin and the admin sidebar.
- It is useful to know that both admins and super_admins will also have access to these pages.
- We also write a little script that gives anyone with tech_admin permissions access to this resource by adding the single_resource_admin role with resource type DataUpdateScript role. 

It may be confusing to you why we also give the user a role single_resource_admin DataUpdateScript when a tech_admin role is selected. To read more about it you can read [this comment](https://github.com/forem/internalEngineering/issues/276#issuecomment-771658292) or the full discussion.

## Related Tickets & Documents
Issue: https://github.com/forem/internalEngineering/issues/276

## QA Instructions, Screenshots, Recordings

Test that the only single_resource _admins with resource type DataUpdateScript, admins and super _admins have access to data scripts.
- Open a rails console with `rails c`
- Find your user that you're logged in with, we're assuming it's the last user `u = User.last`
- Show all that users roles: `roles = u.roles` 
- Keep note of the roles you have and then destroy them `roles.destroy_all'
- Add a single resource admin role to your user: `u.add_role(:single_resource_admin, DataUpdateScript)` 
- You will now have access to this http://localhost:3000/admin/data_update_scripts.
-
You will also be able to see this resource should you have admin or super_admin roles.

<img width="1680" alt="Screenshot 2021-01-15 at 11 16 25" src="https://user-images.githubusercontent.com/2786819/104707814-93340b00-5725-11eb-9ebe-fdf1835ea2db.png">
<img width="1680" alt="Screenshot 2021-01-15 at 11 33 14" src="https://user-images.githubusercontent.com/2786819/104707825-97602880-5725-11eb-9503-81edf5e34c46.png">

Test that a tech_admin given permissions through Manage Users has access to the page:
 - Go http://localhost:3000/admin/users
 - Click on the user that you want to give a tech admin role too
 - Before continuing it would be best to destroy this users current roles to more effectively test the feature. You can do this 
    - Open a rails console with `rails c`
    - Find find the user and assign the User to u.
    - Show all that users roles: `roles = u.roles` 
    - Keep note of the roles you have and then destroy them `roles.destroy_all'
- Now go back to http://localhost:3000/admin/users, click on the same user, click on `Manage User` 
- Select Tech Admin from the dropdown: 
<img width="1575" alt="Screenshot 2021-02-05 at 11 43 10" src="https://user-images.githubusercontent.com/2786819/107017923-792b9c80-67a8-11eb-85cf-07df989113f4.png">
- You should see two roles pop up for the user:
<img width="1051" alt="Screenshot 2021-02-05 at 11 42 33" src="https://user-images.githubusercontent.com/2786819/107018064-a24c2d00-67a8-11eb-9174-b773b40af00f.png">

The user will now have access to http://localhost:3000/admin/data_update_scripts

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
Not as yet.

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/26wkRxKJ9yUZzlorK/giphy.gif)
